File: 01_Harvard_CS50x/Week1/pset1/credit/credit.c
// Author: Jesus Carlos Martinez Gonzalez
// Date: 03/10/2021
// Objective: 

#include <cs50.h> // For educational purposes
#include <stdio.h> 

// Gets the card number's as input and returns the card type if valid, 'INVALID' if otherwise
string card_type(long int);

int main(void)
{
    long int number = get_long("Number: ");
    
    printf("%s\n", card_type(number));
}

string card_type(long int number)
{
    // Variables declaration
    int size = 0, first_digit = -1, second_digit = -1, checksum = 0, curr = 0, aux = 0;
    
    // Iterates over the number, starting from its last digit
    while (number > 0)
    {
        curr = number % 10;
        second_digit = first_digit;
        first_digit = curr;
        number /= 10;
        
        // If the digit is pair, it will get doubled
        if ((size + 1) % 2 == 0)
        {
            curr *= 2;
            
            // If the number is above 9, its first digit will go to curr, second one to aux
            if (curr >= 10)
            {
                aux = curr % 10;
                curr /= 10;
            }
        }
        
        // Adds to checksum
        checksum += curr + aux;
        aux = 0;
        size++;
    }
    
    // Does card-provider validation
    if (checksum % 10 == 0)
    {
        if ((first_digit == 3 && (second_digit == 4 || second_digit == 7)) && size == 15)
        {
            return "AMEX";
        }
        else if ((first_digit == 5 && (second_digit >= 1 && second_digit <= 5)) && size == 16)
        {
            return "MASTERCARD";
        }
        else if (first_digit == 4 && (size == 13 || size == 16))
        {
            return "VISA";
        }
    }
    
    // If validations fail
    return "INVALID";
}

File: 01_Harvard_CS50x/Week6/pset6/credit/credit.py
# Author: Jesus Carlos Martinez Gonzalez
# Date: 23/10/2021
# Objective: Print whether the input number corresponds to a Visa, Mastercard or American Express card


# Performs the checksum validation over the passed argument
def check_sum(number):
    total = 0
    pos = 1
    curr = 0
    for i in reversed(number):
        curr = int(i)
        if pos % 2 == 0:
            curr *= 2
        for j in str(curr):
            total += int(j)
        pos += 1
    if total % 10 == 0:
        return True
    return False


# Validates the card given the argument
def card_type(number):
    if check_sum(number):
        if len(number) == 15 and number[0:2] in ['35', '37']:
            return('AMEX')
        elif len(number) == 16 and number[0:2] in ['51', '52', '53', '54', '55']:
            return('MASTERCARD')
        elif (len(number) == 13 or len(number) == 16) and number[0] == '4':
            return('VISA')
        else: 
            return('INVALID')
    else:
        return('INVALID')
    

def main():
    while True:
        number = input('Enter the card number: ')
        if number.isnumeric():
            break
    print(card_type(number))


if __name__ == "__main__":
    main()